src/librustc_resolve/late.rs

```diff
@@ -556,25 +564,40 @@ impl<'a, 'tcx> Visitor<'tcx> for LateResolutionVisitor<'a, '_> {
                             )
                             .is_some()
                         };
-
-                        if !check_ns(TypeNS) && check_ns(ValueNS) {
-                            // This must be equivalent to `visit_anon_const`, but we cannot call it
-                            // directly due to visitor lifetimes so we have to copy-paste some code.
-                            self.with_constant_rib(|this| {
-                                this.smart_resolve_path(
-                                    ty.id,
-                                    qself.as_ref(),
-                                    path,
-                                    PathSource::Expr(None),
-                                );
-
-                                if let Some(ref qself) = *qself {
-                                    this.visit_ty(&qself.ty);
-                                }
-                                this.visit_path(path, ty.id);
-                            });
-
-                            return;
+                        let in_type_ns = check_ns(TypeNS);
+                        let in_value_ns = check_ns(ValueNS);
+
+                        match (self.diagnostic_metadata.current_item, in_type_ns, in_value_ns) {
```

Comment:
This is a pretty subtle disambiguation logic for generic parameters.
Could you avoid mixing it with diagnostic code?

Reply:
Or perhaps it's a wrong place for `report_missing_type_error` at all.
I need to check.

Reply:
I'm adding a boolean field carrying info on whether we're currently dealing with type params and add the suggestion if true. Do you feel that is less intrusive?

Reply:
I also thought, why is the generic *argument* position a criterion for suggesting generic parameters?
E.g. why is unresolved `Bar` in `foo::<Bar>` more likely to be a generic parameter than unresolved `Bar` in `let foo: Bar`.
The example with `impl Trait<T>` is convincing, but it's some very special case of generic arguments.

It's quite possible that we can recommend introducing generic parameters for all unresolved single-identifier paths in type namespace (perhaps excluding those starting with a lowercase letter).

Reply:
The point is one of likelihood:

```
impl Trait<T> { // definitely
}
fn foo(t: T) -> T { // maybe?
    panic!()
}
fn foo() -> Inexistent { // maybe?
    panic!()
}
fn foo() {
    let x = T; // unlikely
}
fn foo() {
    let x = Foo<T>; // maybe?
}
```

Using the single uppercase letter name for missing types might work in practice, but I worry about having _too many_ suggestions.